Use alwaysUse24HourFormat when formatting time of day

yjbanov opened the pull request:
Add `alwaysUse24HourFormat` property to `MediaQuery` (currently not wired to `dart:ui`; will do it in a follow-up PR).

Drive-by changes:

- Fixes labels on the 24-hour rings, changing them from 0-11/13-24 to 1-12/13-23 to match Android.
- Makes `TimePickerDialog` public in order to support custom `Localizations` and `MediaQuery` (allows better testing and gives developers more control).

Breaking changes:
- adds an optional `alwaysUse24HourFormat` parameter to some methods of `MaterialLocalizations`. This should only affect _implementations_ of this interface. _Users_ of `MaterialLocalizations` should not be affected.


Fixes https://github.com/flutter/flutter/issues/11994
Fixes #4815

Hixie commented on packages/flutter/lib/src/material/time.dart:
```diff
@@ -185,7 +185,7 @@ HourFormat hourFormat({ @required TimeOfDayFormat of }) {
     case TimeOfDayFormat.h_colon_mm_space_a:
     case TimeOfDayFormat.a_space_h_colon_mm:
-      return HourFormat.h;
+      return window.alwaysUse24HourFormat ? HourFormat.HH : HourFormat.h;
```
this code doesn't register to be told when the flag changes, which seems bad. I would recommend putting this flag in the MediaQueryData object and having the caller read that then pass in the flag to this method.

Hixie commented on packages/flutter/test/material/time_test.dart:
```diff
@@ -0,0 +1,24 @@
+      expect(hourFormat(of: TimeOfDayFormat.a_space_h_colon_mm), HourFormat.HH);
+      ui.window.debugOverrideAlwaysUse24HourFormat(false);
+      expect(hourFormat(of: TimeOfDayFormat.a_space_h_colon_mm), HourFormat.h);
+    });
```
you should also check that changing the state of the flag changes the actual rendered widget

yjbanov committed:
alwaysUse24HourFormat in MediaQuery and time picker

yjbanov committed:
docs; dead code

yjbanov commented:
PTAL:

- Merged with @HansMuller's `flutter_localizations` refactoring.
- Moved the `alwaysUse24HourFormat` flag to `MediaQueryData` class.
- Made a drive-by fix to the ring labels (made then 1-12/13-23 to match Android), and added tests.

Hixie commented on .idea/modules.xml:
```diff
@@ -7,9 +7,10 @@
     </modules>
   </component>
-</project>
+</project>
```
is this expected?

Hixie commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -112,14 +112,23 @@ abstract class MaterialLocalizations {
+  ///
+  /// If [alwaysUse24HourFormat] is true, formats hour using [HourFormat.HH]
+  /// rather than the default for the current locale. This value is usually
+  /// passed from [MediaQueryData.alwaysUse24HourFormat], and it has platform-
```
rather than "and it has", consider "which has", since you're talking about MediaQueryData.alwaysUse24HourFormat itself rather than the value of this argument or the behaviour of this method.

Hixie commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -112,14 +112,23 @@ abstract class MaterialLocalizations {
+  /// rather than the default for the current locale. This value is usually
+  /// passed from [MediaQueryData.alwaysUse24HourFormat], and it has platform-
+  /// specific behavior. For more details, see the documentation for
+  /// [MediaQueryData.alwaysUse24HourFormat].
```
if you feel it is important to link to MediaQueryData.alwaysUse24HourFormat twice, consider the more common "see also" style, as in:

```
  /// See also:
  ///
  ///  * [MediaQueryData.alwaysUse24HourFormat], the documentation for which provides more detail.
```
...or something.

Hixie replied:
(or leave it as you have it now if you think that's clearer)

Hixie commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -263,9 +272,20 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+  String formatHour(TimeOfDay timeOfDay, { bool alwaysUse24HourFormat: false }) {
+    if (alwaysUse24HourFormat)
+      return _formatTwoDigitZeroPad(timeOfDay.hour);
+    else
```
you can remove this else

Hixie commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -263,9 +272,20 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+    if (alwaysUse24HourFormat)
+      return _formatTwoDigitZeroPad(timeOfDay.hour);
+    else
+      return formatDecimal(timeOfDay.hourOfPeriod == 0 ? 12 : timeOfDay.hourOfPeriod);
```
interesting that timeOfDay.hourOfPeriod doesn't do this logic for you.

Hixie commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -263,9 +272,20 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+  String _formatTwoDigitZeroPad(int number) {
+    if (number < 10)
+      return '0$number';
+    else
```
remove return

Hixie commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -263,9 +272,20 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+  /// Formats [number] using two digits, assuming it's in the 0-99 inclusive
+  /// range. Not designed to format values outside this range.
+  String _formatTwoDigitZeroPad(int number) {
+    if (number < 10)
```
assert number is 0..99

Hixie commented on packages/flutter/lib/src/material/time.dart:
```diff
@@ -85,7 +85,10 @@ class TimeOfDay {
-    return localizations.formatTimeOfDay(this);
+    return localizations.formatTimeOfDay(
+      this,
+      alwaysUse24HourFormat: MediaQuery.of(context).alwaysUse24HourFormat,
```
if you cannot handle MediaQuery.of() returning null (which seems reasonable, many material widgets require this) then please add an assert to the build method that calls [debugCheckHasMediaQuery](https://master-docs-flutter-io.firebaseapp.com/flutter/widgets/debugCheckHasMediaQuery.html). (See other widgets for examples of doing this.)

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -239,7 +237,10 @@ class _HourControl extends StatelessWidget {
-      child: new Text(localizations.formatHour(fragmentContext.selectedTime), style: hourStyle),
+      child: new Text(localizations.formatHour(
+        fragmentContext.selectedTime,
+        alwaysUse24HourFormat: MediaQuery.of(context).alwaysUse24HourFormat,
```
ditto

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -687,7 +689,7 @@ const List<TimeOfDay> _amHours = const <TimeOfDay>[
 
 const List<TimeOfDay> _pmHours = const <TimeOfDay>[
-  const TimeOfDay(hour: 12, minute: 0),
+  const TimeOfDay(hour: 0, minute: 0),
```
??

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -986,6 +1000,7 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
   Widget build(BuildContext context) {
     final ThemeData themeData = Theme.of(context);
     final MaterialLocalizations localizations = MaterialLocalizations.of(context);
+    final MediaQueryData media = MediaQuery.of(context);
```
ditto assert

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1005,13 +1020,13 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
     switch (widget.mode) {
       case _TimePickerMode.hour:
         if (widget.is24h) {
```
how does is24h interact with the new flag?

Hixie replied:
It seems weird to be able to have the double ring with the 12h mode or the single ring with 24h mode...

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1005,13 +1020,13 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
-          secondaryOuterLabels = _init24HourOuterRing(theme.accentTextTheme, localizations);
-          primaryInnerLabels = _init24HourInnerRing(theme.textTheme, localizations);
-          secondaryInnerLabels = _init24HourInnerRing(theme.accentTextTheme, localizations);
+          primaryOuterLabels = _init24HourOuterRing(theme.textTheme, localizations, media.alwaysUse24HourFormat);
```
i was a bit confused when reading this because I thought i interpreted "_initFoo" as methods that get called from initState. Maybe rename those buildFoo to make it clearer they are called every build?

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1043,20 +1058,30 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
+/// selected [TimeOfDay] if the user taps the "OK" button, or null if the user
+/// taps the "CANCEL" button. The selected time is reported by calling
+/// [Navigator.pop].
+class TimePickerDialog extends StatefulWidget {
```
do we only expose this for testing?

Hixie replied:
seems weird to expose a TimePickerDialog but not a DatePickerDialog.
seems weird to expose either if they're an implementation detail of showTimePicker et al.

yjbanov replied:
Done (used custom Navigator/MediaQuery/Localizations/Directionality).

Hixie commented on packages/flutter/lib/src/widgets/media_query.dart:
```diff
@@ -77,6 +79,19 @@ class MediaQueryData {
+  ///   "Use 24-hour format". It applies to any locale used by the application,
+  ///   whether it is the system-wide locale, or the custom locale set by the
+  ///   application.
+  /// - On iOS this flag is set to `true` when the user setting called "24-Hour
```
avoid backticks around true, false, and null.

Hixie commented on packages/flutter_localizations/flutter_localizations.iml:
```diff
@@ -0,0 +1,14 @@
+    <orderEntry type="library" name="Dart SDK" level="project" />
+    <orderEntry type="library" name="Dart Packages" level="project" />
+  </component>
+</module>
```
see earlier question

Hixie commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -205,4 +204,54 @@ void main() {
+    await tester.pumpWidget(new MaterialApp(
+      home: const MediaQuery(
+        data: const MediaQueryData(alwaysUse24HourFormat: false),
+        child: const TimePickerDialog(initialTime: const TimeOfDay(hour: 7, minute: 0)),
```
you can write this test with showTimePicker, use a GlobalKey to get the context from inside the MediaQuery.

yjbanov replied:
Done (used custom Navigator/MediaQuery/Localizations/Directionality).

Hixie commented:
![LGTM](http://junkyard.damowmow.com/579) modulo comments.

I'd rather we didn't expose TimePickerDialog (just because it will confuse people who want to know how to show a time picker), but if we do expose it, we should definitely expose DatePickerDialog as well and we should thoroughly document both, in particular pointing them to the showTimePicker et al functions, probably with `## Sample code` to show how to use them.

renefloor commented:
I just updated flutter last week (I can check to which version later) and now I have this bug. Can it be I got half of this feature? My device is English and I want the 24h format. I always got the 12h format, but now I get a mix. It would also be good to document how to format timestamps in other textviews.
![timepicker24](https://user-images.githubusercontent.com/15101411/32044874-0d65f03c-ba3f-11e7-9bf0-26acc7378a97.png)

yjbanov committed:
address some comments

yjbanov committed:
MaterialLocalizations.timeOfDayFormat is the single source of 24-hour-formattedness

yjbanov committed:
Make TimePickerDialog private again

yjbanov commented:
@renefloor this PR should fix that too. There was a regression when we moved our default localization out into a `flutter_localizations`, which I discovered while working on this PR (the fix is [here](https://github.com/flutter/flutter/pull/12517/files#diff-d7e5c9c98ba1756f53f3c62c9307fb04R285) if you're curious about the details).

yjbanov commented:
@Hixie did you want to do another pass over the PR or is your LGTM still good? The biggest change after your LGTM was https://github.com/flutter/flutter/pull/12517/commits/341143d37f5c2818f8cbe2397f0992037fdcd947 where `MaterialLocalizations.timeOfDayFormat` becomes the single source of truth for how we format hours.

Hixie commented:
Beautiful. LGTM.

yjbanov committed:
wire up MediaQueryData.fromWindow to Window
